File: packages/depz-sensor-sdk-ts/src/discovery.ts
/**
 * Device discovery & selection (contract 02 §4).
 *
 * The engine here is transport-agnostic: it takes an enumerated list of
 * candidate ports plus a factory that turns a port into a (not-yet-open)
 * `SerialTransport`. The Node entry (`@depz/sensor-sdk/node`) wires it to
 * `serialport`; the browser entry (`@depz/sensor-sdk/web`) wires it to
 * granted WebSerial ports. Both expose `openDevice` / `listDepzDevices`.
 *
 * Selection is by USB identity (`usb-ids.ts`) — never "first port in system
 * order", which could grab an unrelated `/dev/ttyS0`. Mirrors the Python
 * reference `depz_sensor_sdk.discovery`.
 */

import { DepzDevice } from "./device/device.js";
import { DepzError, NoDepzDeviceError } from "./errors.js";
import { parseSoftwareName, type SensorType } from "./protocol/identity.js";
import { isKnownDepzUsb, usbModelHint } from "./protocol/usb-ids.js";
import { Bno086 } from "./sensors/bno086/bno086.js";
import { Sr04 } from "./sensors/sr04.js";
import { Vl53l4Cd } from "./sensors/vl53l4/vl53l4.js";
import { Vl53l8Ch, Vl53l8Cx } from "./sensors/vl53l8/vl53l8.js";
import type { SerialTransport } from "./transport/types.js";

/** One enumerated candidate port (Node: serialport.list; Web: granted port). */
export interface DepzPortInfo {
  /** Stable identifier: an OS path (Node) or a synthetic id (WebSerial). */
  path: string;
  /**
   * USB iSerial as reported by the OS — the sort/selection key. WebSerial
   * cannot read this, so it is undefined there (selection falls back to
   * granted-port order).
   */
  serialNumber?: string;
  usbVid?: number;
  usbPid?: number;
}

/** Result of probing a port (parity with Python `DeviceInfo`). */
export interface DeviceInfo {
  path: string;
  mode: "app" | "bootloader" | "unknown";
  sensorType: SensorType | null;
  softwareName: string;
  fwVersion: string;
  deviceName: string;
  /** Device-reported serial (GET_SERIAL); "" when the device does not answer. */
  serialNumber: string;
  usbVid: number | null;
  usbPid: number | null;
}

/** Turns an enumerated port into a fresh, not-yet-open transport. */
export type TransportFactory = (port: DepzPortInfo) => SerialTransport;

export interface OpenDeviceOptions {
  /** Select the candidate whose USB iSerial equals this exact string. */
  serial?: string;
  /** Per-request timeout for the probe and the opened device (ms). */
  timeoutMs?: number;
}

export interface ListOptions {
  /** Probe only ports with a known DEPZ (vid,pid) (default). */
  matchUsb?: boolean;
  timeoutMs?: number;
}

/** `undefined` = auto; number = Nth candidate; string = port path; DeviceInfo = its port. */
export type OpenTarget = number | string | DeviceInfo | null | undefined;

const DEFAULT_PROBE_TIMEOUT_MS = 200;

// ── selection helpers ─────────────────────────────────────────────────────────

function isKnownPort(p: DepzPortInfo): boolean {
  return isKnownDepzUsb(p.usbVid, p.usbPid);
}

/**
 * Sort ports by USB iSerial ascending; empty/missing serials sort last, ties
 * broken by path. Pure and deterministic — a given bench always maps the same
 * serial to index 0. Exposed so the ordering rule can be golden-tested.
 */
export function orderPortsBySerial(ports: DepzPortInfo[]): DepzPortInfo[] {
  return ports.slice().sort((a, b) => {
    const sa = a.serialNumber ?? "";
    const sb = b.serialNumber ?? "";
    const ea = sa === "" ? 1 : 0;
    const eb = sb === "" ? 1 : 0;
    if (ea !== eb) return ea - eb;
    if (sa !== sb) return sa < sb ? -1 : 1;
    return a.path < b.path ? -1 : a.path > b.path ? 1 : 0;
  });
}

/**
 * Candidate ports (known DEPZ vid/pid) sorted by USB iSerial (index 0 = the
 * one auto-select opens).
 */
function sortedCandidates(ports: DepzPortInfo[]): DepzPortInfo[] {
  return orderPortsBySerial(ports.filter(isKnownPort));
}

export function isDeviceInfo(t: unknown): t is DeviceInfo {
  return typeof t === "object" && t !== null && "path" in t && "sensorType" in t;
}

/**
 * Order *probed* devices by their DEVICE-reported serial (GET_SERIAL, i.e.
 * `DeviceInfo.serialNumber`); empty serials sort last, ties broken by path.
 * This is the selection key for transports that cannot read the USB iSerial
 * (WebSerial), where the serial is only knowable by probing.
 */
export function orderDevicesBySerial(infos: DeviceInfo[]): DeviceInfo[] {
  return infos.slice().sort((a, b) => {
    const ea = a.serialNumber === "" ? 1 : 0;
    const eb = b.serialNumber === "" ? 1 : 0;
    if (ea !== eb) return ea - eb;
    if (a.serialNumber !== b.serialNumber) return a.serialNumber < b.serialNumber ? -1 : 1;
    return a.path < b.path ? -1 : a.path > b.path ? 1 : 0;
  });
}

// ── probing ───────────────────────────────────────────────────────────────────

/**
 * Open `port`, ask its identity (+ device name/serial), then close. Returns
 * null when nothing DEPZ-shaped answers. Probing *opens* the port, so callers
 * should restrict the port list to plausible candidates.
 */
export async function probePort(
  port: DepzPortInfo,
  factory: TransportFactory,
  timeoutMs = DEFAULT_PROBE_TIMEOUT_MS,
): Promise<DeviceInfo | null> {
  const transport = factory(port);
  let dev: DepzDevice;
  try {
    dev = new DepzDevice(transport, { timeoutMs });
    await dev.open();
  } catch {
    try {
      await transport.close();
    } catch {
      /* already gone */
    }
    return null;
  }
  try {
    const ident = parseSoftwareName(await dev.getSoftwareName());
    let deviceName = "";
    try {
      deviceName = await dev.getDeviceName();
    } catch {
      /* optional */
    }
    let serialNumber = "";
    try {
      serialNumber = await dev.getSerialNumber();
    } catch {
      /* optional */
    }
    return {
      path: port.path,
      mode: ident.mode,
      sensorType: ident.sensorType,
      softwareName: ident.softwareName,
      fwVersion: ident.version,
      deviceName,
      serialNumber,
      usbVid: port.usbVid ?? null,
      usbPid: port.usbPid ?? null,
    };
  } catch {
    return null;
  } finally {
    await dev.close();
  }
}

// ── list ──────────────────────────────────────────────────────────────────────

/**
 * Probe candidate ports and return every DEPZ device found, in selection
 * order (sorted by USB iSerial). With `matchUsb` (default) only ports with a
 * known DEPZ (vid,pid) are probed.
 */
export async function listDepzDevicesFrom(
  ports: DepzPortInfo[],
  factory: TransportFactory,
  opts: ListOptions = {},
): Promise<DeviceInfo[]> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_PROBE_TIMEOUT_MS;
  const matchUsb = opts.matchUsb ?? true;
  const scan = matchUsb ? sortedCandidates(ports) : ports.slice();
  const found: DeviceInfo[] = [];
  for (const port of scan) {
    const info = await probePort(port, factory, timeoutMs);
    if (info !== null) found.push(info);
  }
  return found;
}

// ── open ──────────────────────────────────────────────────────────────────────

function constructSensor(
  transport: SerialTransport,
  sensorType: SensorType | null,
  timeoutMs: number,
  port?: DepzPortInfo,
): DepzDevice {
  switch (sensorType) {
    case "sr04":
      return new Sr04(transport, { timeoutMs });
    case "vl53l8":
      // The USB model hint distinguishes CH silicon (PID 0xED40 → 'vl53l8ch')
      // from CX (PID 0xED4B → 'vl53l8cx'); the dev-default id (0x56DC) carries
      // no silicon hint, so it and any non-CH hint fall through to the CX base
      // class.
      return usbModelHint(port?.usbVid, port?.usbPid) === "vl53l8ch"
        ? new Vl53l8Ch(transport, { timeoutMs })
        : new Vl53l8Cx(transport, { timeoutMs });
    case "vl53l4":
      return new Vl53l4Cd(transport, { timeoutMs });
    case "bno086":
      return new Bno086(transport, { timeoutMs });
    default:
      return new DepzDevice(transport, { timeoutMs });
  }
}

async function openProbed(
  port: DepzPortInfo,
  info: DeviceInfo,
  factory: TransportFactory,
  timeoutMs: number,
): Promise<DepzDevice> {
  if (info.mode === "bootloader") {
    throw new DepzError("device is in bootloader mode; use the bootloader client (contract 06)");
  }
  const transport = factory(port);
  const dev = constructSensor(transport, info.sensorType, timeoutMs, port);
  await dev.open();
  return dev;
}

/**
 * Open the right sensor class for `target` (contract 02 §4).
 *
 * - `undefined`/`null`: the candidate with the alphabetically-smallest USB
 *   iSerial (or the one whose serial == `opts.serial`). No candidate → throws
 *   `NoDepzDeviceError` (never falls back to "first port in system").
 * - `number N`: the Nth candidate, sorted by USB iSerial (0-based). Out of
 *   range → `NoDepzDeviceError`.
 * - `string`: open exactly that port path; warns if its (vid,pid) is not a
 *   known DEPZ id (may be an unprogrammed unit).
 * - `DeviceInfo`: opens its `.path`.
 */
export async function openDeviceFrom(
  ports: DepzPortInfo[],
  factory: TransportFactory,
  target?: OpenTarget,
  opts: OpenDeviceOptions = {},
): Promise<DepzDevice> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_PROBE_TIMEOUT_MS;

  // Explicit port path (string or DeviceInfo) — open exactly that port.
  if (typeof target === "string" || isDeviceInfo(target)) {
    const path = typeof target === "string" ? target : target.path;
    const enumerated = ports.find((p) => p.path === path);
    const port: DepzPortInfo = enumerated ?? { path };
    const info = await probePort(port, factory, timeoutMs);
    if (info === null) {
      throw new NoDepzDeviceError(`no DEPZ device answered on ${path}`);
    }
    if (!isKnownDepzUsb(port.usbVid, port.usbPid)) {
      console.warn(
        `depz: ${path} has an unrecognized USB id ` +
          `(vid=${fmtId(port.usbVid)}, pid=${fmtId(port.usbPid)}); opening anyway`,
      );
    }
    if (opts.serial !== undefined && enumerated?.serialNumber !== undefined && enumerated.serialNumber !== opts.serial) {
      console.warn(
        `depz: requested serial ${opts.serial} but ${path} enumerates as ` +
          `${enumerated.serialNumber}`,
      );
    }
    return openProbed(port, info, factory, timeoutMs);
  }

  const candidates = sortedCandidates(ports);

  // Nth candidate by USB iSerial.
  if (typeof target === "number") {
    if (!Number.isInteger(target) || target < 0 || target >= candidates.length) {
      throw new NoDepzDeviceError(
        `no DEPZ device at index ${target} (${candidates.length} candidate(s) found)`,
      );
    }
    const port = candidates[target]!;
    if (opts.serial !== undefined && port.serialNumber !== opts.serial) {
      console.warn(
        `depz: index ${target} selects serial ${port.serialNumber ?? "(none)"}, ` +
          `not the requested ${opts.serial}`,
      );
    }
    return openIndexed(port, factory, timeoutMs);
  }

  // Auto: by serial, else smallest serial (index 0).
  let port: DepzPortInfo | undefined;
  if (opts.serial !== undefined) {
    port = candidates.find((p) => p.serialNumber === opts.serial);
    if (port === undefined) {
      throw new NoDepzDeviceError(`no DEPZ device with serial ${opts.serial}`);
    }
  } else {
    port = candidates[0];
    if (port === undefined) {
      throw new NoDepzDeviceError("no DEPZ device found (no port with a known DEPZ USB id)");
    }
  }
  return openIndexed(port, factory, timeoutMs);
}

/**
 * Select and open a device by its DEVICE-reported serial (GET_SERIAL), for
 * transports that cannot enumerate USB iSerials (WebSerial). Probes every
 * given port (open → GET_SERIAL → close), orders the answers by device
 * serial, then selects — `opts.serial` picks an exact match, a numeric
 * `target` picks the Nth, otherwise the smallest serial — and reopens it.
 *
 * The probing cost (opening each granted port once) is the price of not
 * having USB iSerials up front; the *result* matches Node's serial-based
 * selection because for programmed units USB iSerial == device serial.
 */
export async function openDeviceByDeviceSerial(
  ports: DepzPortInfo[],
  factory: TransportFactory,
  target?: number | null,
  opts: OpenDeviceOptions = {},
): Promise<DepzDevice> {
  const timeoutMs = opts.timeoutMs ?? DEFAULT_PROBE_TIMEOUT_MS;
  // matchUsb:false — a granted port is user-curated and may not expose a
  // recognizable vid/pid; the protocol probe is the real filter.
  const infos = orderDevicesBySerial(
    await listDepzDevicesFrom(ports, factory, { matchUsb: false, timeoutMs }),
  );
  let chosen: DeviceInfo | undefined;
  if (opts.serial !== undefined) {
    chosen = infos.find((d) => d.serialNumber === opts.serial);
    if (chosen === undefined) throw new NoDepzDeviceError(`no DEPZ device with serial ${opts.serial}`);
  } else if (typeof target === "number") {
    if (!Number.isInteger(target) || target < 0 || target >= infos.length) {
      throw new NoDepzDeviceError(
        `no DEPZ device at index ${target} (${infos.length} candidate(s) found)`,
      );
    }
    chosen = infos[target]!;
  } else {
    chosen = infos[0];
    if (chosen === undefined) throw new NoDepzDeviceError("no DEPZ device found");
  }
  const port = ports.find((p) => p.path === chosen!.path) ?? { path: chosen!.path };
  return openProbed(port, chosen, factory, timeoutMs);
}

async function openIndexed(
  port: DepzPortInfo,
  factory: TransportFactory,
  timeoutMs: number,
): Promise<DepzDevice> {
  const info = await probePort(port, factory, timeoutMs);
  if (info === null) {
    throw new NoDepzDeviceError(`no DEPZ device answered on ${port.path}`);
  }
  return openProbed(port, info, factory, timeoutMs);
}

function fmtId(v: number | null | undefined): string {
  return v == null ? "?" : `0x${v.toString(16).toUpperCase().padStart(4, "0")}`;
}

File: packages/depz-sensor-sdk-python/depz_sensor_sdk/discovery.py
"""Device discovery: pick the right DEPZ serial port by USB identity, then
probe it (contract 02 §4).

Default selection considers only ports whose USB (vid, pid) is a known DEPZ
identity (`usb_ids`), ordered by USB iSerial — never "first port the OS
enumerated". The protocol probe (GET_NAME_ACTIVE_SOFTWARE) still decides what a
device actually is; the USB table only chooses which port(s) to talk to.
"""

from __future__ import annotations

import warnings
from dataclasses import dataclass
from typing import Literal

from serial.tools import list_ports

from .device import DeviceBase
from .errors import DepzError, DepzTimeoutError, DeviceLostError, NoDepzDeviceError
from .protocol.identity import Identity, SensorType, parse_software_name
from .sr04 import Sr04
from .transport.link import Link
from .usb_ids import is_known_depz_usb, usb_model_hint


@dataclass(frozen=True)
class DeviceInfo:
    port: str
    mode: Literal["app", "bootloader", "unknown"]
    sensor_type: SensorType | None
    software_name: str
    fw_version: str
    device_name: str
    serial_number: str  # device protocol serial (GET_SERIAL)
    usb_vid: int | None = None  # informative only — never filtered on
    usb_pid: int | None = None
    usb_serial: str | None = None  # USB iSerial, used for ordering/selection

    @property
    def usb_model_hint(self) -> str | None:
        return usb_model_hint(self.usb_vid, self.usb_pid)


@dataclass(frozen=True)
class _PortId:
    """One enumerated serial port with its USB metadata."""

    port: str
    vid: int | None
    pid: int | None
    usb_serial: str | None


def _enumerate_ports() -> list[_PortId]:
    return [
        _PortId(p.device, p.vid, p.pid, p.serial_number)
        for p in list_ports.comports()
    ]


def _serial_sort_key(pi: _PortId) -> tuple[int, str, str]:
    """Sort by USB serial; None/empty sorts last, tie-break by port path."""
    if pi.usb_serial:
        return (0, pi.usb_serial, pi.port)
    return (1, "", pi.port)


def _depz_candidates() -> list[_PortId]:
    """Enumerated ports with a known DEPZ USB identity, ordered by serial."""
    cands = [pi for pi in _enumerate_ports() if is_known_depz_usb(pi.vid, pi.pid)]
    cands.sort(key=_serial_sort_key)
    return cands


def probe_port(port: str, *, timeout: float = 0.2) -> DeviceInfo | None:
    """Open `port`, ask GET_NAME_ACTIVE_SOFTWARE (+ name/serial), close.

    Returns None when nothing DEPZ-shaped answers. Note: probing opens the
    port — skip ports owned by other software via the `ports=` argument of
    `list_depz_devices`.
    """
    try:
        dev = DeviceBase(port, timeout=timeout)
    except (DeviceLostError, DepzError):
        return None
    try:
        software = dev.get_software_name()
        ident = parse_software_name(software)
        try:
            device_name = dev.get_device_name()
        except (DepzTimeoutError, DepzError):
            device_name = ""
        try:
            serial_number = dev.get_serial_number()
        except (DepzTimeoutError, DepzError):
            serial_number = ""
        return DeviceInfo(
            port=port,
            mode=ident.mode,  # type: ignore[arg-type]
            sensor_type=ident.sensor_type,
            software_name=ident.software_name,
            fw_version=ident.version,
            device_name=device_name,
            serial_number=serial_number,
        )
    except (DepzTimeoutError, DeviceLostError, DepzError):
        return None
    finally:
        dev.close()


def list_depz_devices(
    *, ports: list[str] | None = None, timeout: float = 0.2, match_usb: bool = True
) -> list[DeviceInfo]:
    """Probe candidate serial ports and return every DEPZ device found.

    With `match_usb` (default) only ports whose USB (vid, pid) is a known DEPZ
    identity are probed — fast, and it avoids poking unrelated devices. Pass
    `match_usb=False` for the legacy probe-every-port behavior. `ports`
    overrides enumeration with an explicit list (still USB-annotated when the
    OS knows the port). Results are ordered by USB iSerial.
    """
    enumerated = {pi.port: pi for pi in _enumerate_ports()}
    if ports is None:
        candidates = list(enumerated.values())
        if match_usb:
            candidates = [pi for pi in candidates if is_known_depz_usb(pi.vid, pi.pid)]
    else:
        candidates = [
            enumerated.get(p, _PortId(p, None, None, None)) for p in ports
        ]
        if match_usb:
            candidates = [pi for pi in candidates if is_known_depz_usb(pi.vid, pi.pid)]
    candidates.sort(key=_serial_sort_key)

    found: list[DeviceInfo] = []
    for pi in candidates:
        info = probe_port(pi.port, timeout=timeout)
        if info is not None:
            found.append(
                DeviceInfo(
                    **{
                        **info.__dict__,
                        "usb_vid": pi.vid,
                        "usb_pid": pi.pid,
                        "usb_serial": pi.usb_serial,
                    }  # type: ignore[arg-type]
                )
            )
    return found


def open_device(
    target: str | int | DeviceInfo | Link | None = None,
    *,
    serial: str | None = None,
    timeout: float = 0.2,
):
    """Open the right sensor class for `target`.

    - ``None`` (default): the DEPZ candidate port with the alphabetically
      smallest USB serial (index 0); with ``serial=`` the candidate whose USB
      serial matches. No candidate → `NoDepzDeviceError`.
    - ``int N``: the Nth DEPZ candidate, candidates sorted by USB serial
      (0-based). Out of range → `NoDepzDeviceError`.
    - ``str`` (a port path): open exactly that port; if its USB (vid, pid) is
      not a known DEPZ id, warn but proceed.
    - ``DeviceInfo``: its ``.port``.
    - ``Link``/transport: opened directly (loopback/replay).

    Bootloader-mode devices raise until the bootloader client lands
    (contract 06).
    """
    if isinstance(target, Link):
        dev = DeviceBase(target, timeout=timeout)
        ident = _identify(dev)
        return _promote(dev, ident)

    if isinstance(target, DeviceInfo):
        if serial is not None and target.usb_serial not in (None, serial):
            warnings.warn(
                f"requested serial {serial!r} but DeviceInfo serial is "
                f"{target.usb_serial!r}",
                stacklevel=2,
            )
        return _open_port(
            target.port,
            target.sensor_type,
            target.mode,
            timeout,
            usb_model_hint(target.usb_vid, target.usb_pid),
        )

    if isinstance(target, bool):  # guard: bool is an int subclass
        raise TypeError("open_device target must be None, int, str, DeviceInfo or Link")

    if isinstance(target, int):
        candidates = _depz_candidates()
        if serial is not None:
            candidates = [pi for pi in candidates if pi.usb_serial == serial]
        if not 0 <= target < len(candidates):
            raise NoDepzDeviceError(
                f"no DEPZ candidate at index {target} "
                f"({len(candidates)} candidate(s) found)"
            )
        return _open_probed(candidates[target].port, timeout)

    if isinstance(target, str):
        enumerated = {pi.port: pi for pi in _enumerate_ports()}
        pi = enumerated.get(target)
        if pi is not None and not is_known_depz_usb(pi.vid, pi.pid):
            warnings.warn(
                f"{target}: USB id "
                f"{_fmt_vidpid(pi.vid, pi.pid)} is not a known DEPZ device — "
                "opening anyway (unprogrammed unit / custom setup?)",
                stacklevel=2,
            )
        if serial is not None and pi is not None and pi.usb_serial not in (None, serial):
            warnings.warn(
                f"{target}: USB serial {pi.usb_serial!r} != requested {serial!r}",
                stacklevel=2,
            )
        return _open_probed(target, timeout)

    # target is None: select a candidate by serial / smallest-serial default.
    candidates = _depz_candidates()
    if serial is not None:
        match = [pi for pi in candidates if pi.usb_serial == serial]
        if not match:
            raise NoDepzDeviceError(f"no DEPZ device with USB serial {serial!r}")
        return _open_probed(match[0].port, timeout)
    if not candidates:
        raise NoDepzDeviceError(
            "no DEPZ device found (no serial port has a known DEPZ USB id). "
            "Pass a port path explicitly to open an unprogrammed unit."
        )
    return _open_probed(candidates[0].port, timeout)


def _fmt_vidpid(vid: int | None, pid: int | None) -> str:
    if vid is None or pid is None:
        return "unknown"
    return f"{vid:04X}:{pid:04X}"


def _open_probed(port: str, timeout: float):
    info = probe_port(port, timeout=timeout)
    if info is None:
        raise DeviceLostError(f"no DEPZ device answered on {port}")
    pi = {p.port: p for p in _enumerate_ports()}.get(port)
    model = usb_model_hint(pi.vid, pi.pid) if pi is not None else None
    return _open_port(port, info.sensor_type, info.mode, timeout, model)


def _open_port(
    port: str,
    sensor: SensorType | None,
    mode: str,
    timeout: float,
    usb_model: str | None = None,
):
    if mode == "bootloader":
        raise DepzError("device is in bootloader mode; use the bootloader client (M2)")
    return _open_by_sensor(port, sensor, timeout, usb_model)


def _identify(dev: DeviceBase) -> Identity:
    return parse_software_name(dev.get_software_name())


def _open_by_sensor(
    port: str, sensor: SensorType | None, timeout: float, usb_model: str | None = None
):
    if sensor == SensorType.SR04:
        return Sr04(port, timeout=timeout)
    if sensor == SensorType.VL53L8:
        # CX and CH both report firmware name APP_VL53L8_v*, and the silicon
        # device_id/revision is identical (0xF0/0x0C), so the USB PID is the
        # only thing that tells them apart: 0xED4B → CX, 0xED40 → CH (both
        # ship; both are hw-verified in usb_ids.py). A dev unit carrying the
        # STMicro default 0x56DC can't be identified at all, so it falls back
        # to the CX base — the safe choice, since CH is a strict superset.
        from .vl53l8 import Vl53l8Ch, Vl53l8Cx

        cls = Vl53l8Ch if usb_model == "vl53l8ch" else Vl53l8Cx
        return cls(port, timeout=timeout)
    if sensor == SensorType.VL53L4:
        from .vl53l4 import Vl53l4Cd

        return Vl53l4Cd(port, timeout=timeout)
    if sensor == SensorType.BNO086:
        from .bno086 import Bno086

        return Bno086(port, timeout=timeout)
    return DeviceBase(port, timeout=timeout)


def _promote(dev: DeviceBase, ident: Identity):
    """Re-wrap an already-open Link-based DeviceBase as the right subclass.

    State first, class swap second: the reader thread may dispatch a report
    between the two statements."""
    if ident.sensor_type == SensorType.SR04:
        Sr04._init_subclass_state(dev)  # type: ignore[arg-type]
        dev.__class__ = Sr04
        return dev
    if ident.sensor_type == SensorType.VL53L8:
        from .vl53l8 import Vl53l8

        Vl53l8._init_subclass_state(dev)  # type: ignore[arg-type]
        dev.__class__ = Vl53l8
        return dev
    if ident.sensor_type == SensorType.VL53L4:
        from .vl53l4 import Vl53l4Cd

        Vl53l4Cd._init_subclass_state(dev)  # type: ignore[arg-type]
        dev.__class__ = Vl53l4Cd
        return dev
    if ident.sensor_type == SensorType.BNO086:
        from .bno086 import Bno086

        Bno086._init_subclass_state(dev)  # type: ignore[arg-type]
        dev.__class__ = Bno086
        return dev
    return dev
